reps/2022-04-07-serialization.md

```diff
@@ -0,0 +1,162 @@
+## Summary
+### General Motivation
+Current Ray's serialization has some issues:
+
+1. Doesn't support out-of-band serialization in Java worker. Cause some perfomance issues in Java worker.
+1. Hard to add a new serializer for a specific class.
+1. Doens't support cross-language serialization for custom classes.
+
+In order to resolve the above issues. We propose to refactor the current serialization code path, provide a standard way for users to develop their own serializers. 
+In conculution:
+
+1. Provide pluggable ways for users to implement custom serialization, including:
+   1. Cross-language serialization.
+   1. Out of band serialization and other optimizations.
+2. Unify the current serialization code path to this new pluggable design. Make code cleaner.
+### Should this change be within `ray` or outside?
+It should be a part of the main `ray` project.
+## Stewardship
+### Required Reviewers
+Hao Chen, Qing Wang, Eric Liang, Zhi Lin, Simon Mo
+### Shepherd of the Proposal (should be a senior committer)
+Hao Chen
+## Design and Architecture
+### User-level API
+#### Register Serializer
+Firstly, if a user wants to implement custom serialization, he should register his serializer to Ray.
+Note that a unique string ID should be provided for class identification in cross-language serialization.
+```java
+// In Python
+Ray.register_serializer("ArrowTable", type(arrow_table_obj), ArrowTableSerializer());
+Ray.register_serializer("Protobuf", type(protobuf_obj), ProtobufSerializer());
+
+// In Java
+Ray.registerSerializer("ArrowTable", ArrowTable.class, new ArrowTableSerializer());
+Ray.registerSerializer("Protobuf", Protobuf.class, new ProtobufSerializer());
+```
+Ray will maintain a map from ClassID to Serializer in memory. This relationship won't be persisted, which means you need to register them again when the process is restarted.
+#### Implement Serializer
+Then, he needs to implement this custom serializer, implements the interface shown below:
+Let's take Python as an example. Other languages will be similar.
+```python
+class MySerializer(RaySerializer):
+    def serialize(self, object: MyClass) -> RaySerializationResult:
+        pass
+    def deserialize(self, in_band_data: bytes,
+                    out_of_band_data: List[bytes]) -> MyClass:
+        pass
+
+```
+`serialize`method should return a `SerializedResult`, which contains 2 fields: in-band buffer and out-of-band buffers.
+`inBandBuffer` is nothing else than a byte array. Users can use this field to achieve normal in-band serialization.
+```python
+class RaySerializationResult:
+    def __init__(self):
+        self.in_band_buffer: bytes = None
```

Comment:
maybe also a list of Ray objectrefs? Same in the deserialization API.

Reply:
Why Ray object refs? I can't get your point. Could you please say more about this?

Reply:
I think this is maybe out of context and orthogonal for this REP. Let me resolve this.